### DISCUSSION TITLE
Publish MQTT when ENERGY drops to 0 and POWER>1

### DISCUSSION THREAD

**Discussion by patrmis** [arendst/Tasmota#18787] [Support request] [answered]
I have installed tasmota 12.5.0 on a Shelly 2.5, enabled shutter support and configured as I needed (time limits etc).
My application (ventilation damper) has an end switch which cuts the power to the motor once the limit (open or closed) has been reached, and I'm really only interested in fully open or fully closed, hence the Shutter[Close/Open]Duration is not critical.

I would like to publish an 'OPENING', 'CLOSING', 'OPEN' and 'CLOSED' message to MQTT when the shutter is running, and when it's still running but the measured power drops to zero i.e. end limit has been reached, but shutter run duration not yet reached.

My attempt at the rules are:
`Rule1
on POWER1#STATE=1 do publish cmnd/%topic%/POSITION OPENING endon
on POWER2#STATE=1 do publish cmnd/%topic%/POSITION CLOSING endon`

`Rule2
on POSITION$!CLOSING do var1 1 break
on ENERGY#Power[2]<1 do publish cmnd/%topic%/POSITION OPEN endon`

`Rule3
on POSITION$!OPENING do var1 1 break
on ENERGY#Power[1]<1 do publish cmnd/%topic%/POSITION CLOSED endon`

However rule2 and rule3 seem to just spam the console with `{"Command":"Unknown"}` and MQTT with 'POSITION OPEN' and 'POSITION CLOSED' commands.
From this I would propose the first line of the rule2 and 3 don't have the intended effect (i.e. do nothing but BREAK when the POSITION command is **not** 'OPENING'/'CLOSING'.

What would be the appropriate rule2 and rule3 for the intended purpose?

For info, the tele transmission is:
`13:03:57.152 MQT: tele/vent_office_zone/STATE = {"Time":"2023-06-04T13:03:57","Uptime":"0T00:15:09","UptimeSec":909,"Heap":24,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"POWER1":"OFF","POWER2":"OFF","Wifi":{"AP":1,"SSId":"home_automation","BSSId":"B0:4E:26:45:9C:D6","Channel":11,"Mode":"11n","RSSI":100,"Signal":-47,"LinkCount":1,"Downtime":"0T00:00:03"}}`

`13:03:57.168 MQT: tele/vent_office_zone/SENSOR = {"Time":"2023-06-04T13:03:57","Switch1":"OFF","Switch2":"OFF","ANALOG":{"Temperature":60.0},"ENERGY":{"TotalStartTime":"2023-06-03T20:21:51","Total":0.001,"Yesterday":0.000,"Today":0.001,"Period":[0,0],"Power":[0,0],"ApparentPower":[0,0],"ReactivePower":[0,0],"Factor":[0.00,0.00],"Frequency":0,"Voltage":0,"Current":[0.000,0.000]},"Shutter1":{"Position":0,"Direction":0,"Target":0,"Tilt":0},"TempUnit":"C"}`

**Comment by sfromis** [comment]
Do not publish to your own command topic (cmnd). Better publish to stat instead, or any other topic you choose.

The reason why your `break` does not work is that rules are triggered by independent events, and `break` can only work during processing of the same event. A general way around this is to have one rule save the state info into a var, which another rule then can depend on.

**Comment by patrmis** [comment, marked as answer]
Thank you sfromis.

I never did manage to make it work with the `break` statement.
Have instead opted to compile through [TasmoCompiler](https://github.com/benzino77/tasmocompiler) with `USE_EXPRESSION` and  `SUPPORT_IF_STATEMENT`

And have settled on the below rules, which seem to work quite well:

`Rule1`
`ON POWER1#STATE=1 DO backlog  var1 0; ruletimer1 4; publish stat/%topic%/POSITION OPENING; ENDON`
`ON rules#timer=1 DO var1 1 ENDON`
`ON ENERGY#CURRENT[1]<0.01 DO IF (var1==1) var1 0; POWER1 0; POWER2 0; publish stat/%topic%/POSITION OPEN; ENDIF ENDON`
`ON POWER1#STATE=0 DO backlog var1 0; ruletimer1 0; ruletimer2 0 ENDON`

`Rule2`
`ON POWER2#STATE=1 DO backlog var1 0; ruletimer2 4; publish stat/%topic%/POSITION CLOSING; ENDON`
`ON rules#timer=2 DO var1 2 ENDON`
`ON ENERGY#CURRENT[2]<0.01 DO IF (var1==2) var1 0; POWER1 0; POWER2 0; publish stat/%topic%/POSITION CLOSED; ENDIF ENDON`
`ON POWER2#STATE=0 DO backlog var1 0; ruletimer1 0; ruletimer2 0 ENDON`

**Comment by patrmis** [comment]
What happens with the IF statement now though, I get an MQTT message `RESULT = {"If":"Done"}` about twice a second, every second continuously (not just during opening/closing events).

Considering I'll end up with about 10 devices, is there a way to turn the publish of that off to avoid the network spamming?

Is this significant traffic on a (typical 802.11n wifi) network? 
For the Eclipse Mosquitto MQTT server (running on Raspberry Pi 4, 8Gb)?
Or not something I need to worry about, as it's insignificant.

**Reply by sfromis** [reply]
No, Tasmota does not facilitate turning off that **tiny** insignificant amount of traffic.

**Reply by patrmis** [reply]
Point taken. 
Thank you.
